sdks/java/io/thrift/src/main/java/org/apache/beam/sdk/io/thrift/ThriftIO.java

```
@@ -0,0 +1,289 @@
+
+    abstract Builder<T> toBuilder();
+
+    @Nullable
+    abstract Class<T> getRecordClass();
+
+    @Nullable
+    abstract TProtocolFactory getTProtocolFactory();
+
+    public ReadFiles<T> withProtocol(TProtocolFactory protocol) {
+      return toBuilder().setTProtocolFactory(protocol).build();
+    }
+
+    @Override
+    public PCollection<T> expand(PCollection<FileIO.ReadableFile> input) {
+      checkNotNull(getRecordClass(), "Record class cannot be null");
+      checkNotNull(getTProtocolFactory(), "Thrift protocol cannot be null");
+
+      final Coder<T> coder = ThriftCoder.of();
+      return input
+          .apply(ParDo.of(new ReadFn<>(getRecordClass(), getTProtocolFactory())))
+          .setCoder(coder);
+    }
+
+    @AutoValue.Builder
+    abstract static class Builder<T> {
+      abstract Builder<T> setRecordClass(Class<T> recordClass);
+
+      abstract TProtocolFactory getTProtocolFactory();
+
+      abstract Builder<T> setTProtocolFactory(TProtocolFactory tProtocol);
+
+      abstract ReadFiles<T> autoBuild();
+
+      public ReadFiles<T> build() {
+        return autoBuild();
+      }
+    }
+
+    /**
+     * Reads a {@link FileIO.ReadableFile} and outputs a {@link PCollection} of {@link
+     * #getRecordClass()}.
+     */
+    static class ReadFn<T> extends DoFn<FileIO.ReadableFile, T> {
+
+      final Class<T> tBaseType;
+      final TProtocolFactory tProtocol;
+
+      ReadFn(Class<T> tBaseType, TProtocolFactory tProtocol) {
+        this.tBaseType = tBaseType;
+        this.tProtocol = tProtocol;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext processContext) {
+        FileIO.ReadableFile file = processContext.element();
+
+        try {
+
+          InputStream inputStream = Channels.newInputStream(file.open());
+          TIOStreamTransport streamTransport = new TIOStreamTransport(inputStream);
+          TProtocol protocol = tProtocol.getProtocol(streamTransport);
+
+          while (true) {
```

you should be able to check if the stream has more data here each iteration, rather than relying on the EOF exception.  There are other reasons you could get a TTransportException, and this will silently swallow them, resulting in potentially corrupt data.

---

agreed. The next commit will wrap the `TIOStreamTransport` with an  `AutoExpandingBufferReadTransport` so that `protocol.getTransport().getBytesRemainingInBuffer()` can be used to iterate.